meilisearch-lib/src/tasks/handlers/index_resolver_handler.rs

```diff
@@ -0,0 +1,146 @@
+use crate::index_resolver::IndexResolver;
+use crate::index_resolver::{index_store::IndexStore, meta_store::IndexMetaStore};
+use crate::tasks::batch::{Batch, BatchContent};
+use crate::tasks::task::TaskEvent;
+use crate::tasks::BatchHandler;
+
+#[async_trait::async_trait]
+impl<U, I> BatchHandler for IndexResolver<U, I>
+where
+    U: IndexMetaStore + Send + Sync + 'static,
+    I: IndexStore + Send + Sync + 'static,
+{
+    fn accept(&self, batch: &Batch) -> bool {
+        matches!(
+            batch.content,
+            BatchContent::DocumentsAdditionBatch(_) | BatchContent::IndexUpdate(_)
+        )
+    }
+
+    async fn process_batch(&self, mut batch: Batch) -> Batch {
+        match batch.content {
+            BatchContent::DocumentsAdditionBatch(ref mut tasks) => {
+                *tasks = self
+                    .process_document_addition_batch(std::mem::take(tasks))
+                    .await;
+            }
+            BatchContent::IndexUpdate(ref mut task) => match self.process_task(task).await {
+                Ok(success) => task.events.push(TaskEvent::succeeded(success)),
+                Err(err) => task.events.push(TaskEvent::failed(err.into())),
+            },
+            _ => unreachable!(),
+        }
+
+        batch
+    }
```

Comment:
I find it quite strange to have an `accept` method + an `unreachable` in the `process_batch`.
Couldn't we implement a `process(self, Context)` on the tasks directly?

It might be strange too to have a `Context` which contains everything though, but I think you have access to everything when you call the accept to process_batch methods.

I don't think we should change that right now, but I would like to hear if you considered doing it the other way around and what was the issue.

Reply:
On the task or on the batch? What would Context be?

Reply:
> On the task or on the batch? What would Context be?

On the task. The `Context` would be something that contains;
- The `IndexMetaStore`
- The `IndexStore`
- The `SnapshotHandler`
- The `DumpHandler`

Basically, everything that implements `BatchHandler` currently.

Reply:
I'm not convinced about this approach, it doesn't feel right that tasks process themselves to me, they're just data after all. we would be shifting a lot of logic from the handlers to the tasks, and that would not be right for IndexResolver. It seems natural that IndexResolver has the logic necessary to apply an update to an index, hence the logic to do so is in its implementation of `BatchHandler`. While this seems a bit less true for `BatchHandler` because I've just introduced it, for me it contains the logic of performing the batch into a struct whose job is explicitly to do that.

If the goal is to remove the accepted method, I can agree with that and let do what  I proposed, otherwise I am interested in stronger arguments that would justify a change (even though we're not doing it know as you said:) )